accounts/abi/bind/backends/blockchain.go

```diff
@@ -0,0 +1,140 @@
+// Its intended purpose is reading system contracts during block processing.
+//
+// Note that SimulatedBackend creates a new temporary BlockChain for testing,
+// whereas BlockchainContractCaller uses an existing BlockChain with existing database.
+type BlockchainContractCaller struct {
+	bc BlockChainForCaller
+}
+
+// This nil assignment ensures at compile time that BlockchainContractCaller implements bind.ContractCaller.
+var _ bind.ContractCaller = (*BlockchainContractCaller)(nil)
+
+func NewBlockchainContractCaller(bc BlockChainForCaller) *BlockchainContractCaller {
+	return &BlockchainContractCaller{
+		bc: bc,
+	}
+}
+
+func (b *BlockchainContractCaller) CodeAt(ctx context.Context, account common.Address, blockNumber *big.Int) ([]byte, error) {
+	if _, state, err := b.getBlockAndState(blockNumber); err != nil {
+		return nil, err
+	} else {
+		return state.GetCode(account), nil
+	}
+}
+
+// Executes a read-only function call with respect to the specified block's state, or latest state if not specified.
+//
+// Returns call result in []byte.
+// Returns error when:
+// - cannot find the corresponding block or stateDB
+// - VM revert error
+// - VM other errors (e.g. NotProgramAccount, OutOfGas)
+// - Error outside VM
+func (b *BlockchainContractCaller) CallContract(ctx context.Context, call klaytn.CallMsg, blockNumber *big.Int) ([]byte, error) {
+	block, state, err := b.getBlockAndState(blockNumber)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := b.callContract(call, block, state)
+	if err != nil {
+		return nil, err
+	}
+	if len(res.Revert()) > 0 {
+		return nil, newRevertError(res)
+	}
+	return res.Return(), res.Unwrap()
+}
+
+func (b *BlockchainContractCaller) callContract(call klaytn.CallMsg, block *types.Block, state *state.StateDB) (*blockchain.ExecutionResult, error) {
+	if call.Gas == 0 {
+		call.Gas = uint64(3e8) // enough gas for ordinary contract calls
+	}
+
+	intrinsicGas, err := types.IntrinsicGas(call.Data, nil, call.To == nil, b.bc.Config().Rules(block.Number()))
+	if err != nil {
+		return nil, err
+	}
+
+	msg := types.NewMessage(call.From, call.To, 0, call.Value, call.Gas, call.GasPrice, call.Data,
+		false, intrinsicGas)
+
+	evmContext := blockchain.NewEVMContext(msg, block.Header(), b.bc, nil)
+	// EVM demands the sender to have enough KLAY balance (gasPrice * gasLimit) in buyGas()
+	// After KIP-71, gasPrice is nonzero baseFee, regardless of the msg.gasPrice (usually 0)
+	// But our sender (usually 0x0) won't have enough balance. Instead we override gasPrice = 0 here
+	evmContext.GasPrice = big.NewInt(0)
+	evm := vm.NewEVM(evmContext, state, b.bc.Config(), &vm.Config{})
+
+	return blockchain.ApplyMessage(evm, msg)
+}
+
+func (b *BlockchainContractCaller) getBlockAndState(num *big.Int) (*types.Block, *state.StateDB, error) {
+	var header *types.Header
+	if num == nil {
+		header = b.bc.CurrentHeader()
+	} else {
+		header = b.bc.GetHeaderByNumber(num.Uint64())
+	}
```

Comment:
@2dvorak We better change this code to use CurrentBlock.Header instead of CurrentHeader. The currentHeader may have a different value. For more detail, see https://github.com/klaytn/klaytn/pull/1766

Reply:
@blukat29 handled this in #1882

Reply:
@aidan-kwon This will be handled in #1893